core/server/master/src/main/java/alluxio/master/file/AccessTimeUpdater.java

```diff
@@ -0,0 +1,176 @@
+  }
+
+  /**
+   * Constructs a new {@link AccessTimeUpdater} with time configurations.
+   */
+  @VisibleForTesting
+  public AccessTimeUpdater(FileSystemMaster fileSystemMaster, InodeTree inodeTree,
+      JournalSystem journalSystem, long flushInterval, long updatePrecision, long shutdownTimeout) {
+    mFileSystemMaster = fileSystemMaster;
+    mInodeTree = inodeTree;
+    mAccessTimeUpdates = new ConcurrentHashSet<>();
+    mFlushInterval = flushInterval;
+    mUpdatePrecision = updatePrecision;
+    mShutdownTimeout = shutdownTimeout;
+    journalSystem.addJournalSink(mFileSystemMaster, this);
+  }
+
+  public void start() {
+    start(mFlushInterval > 0 ? Executors.newSingleThreadScheduledExecutor(
+        ThreadFactoryUtils.build("AccessTimeUpdater-%d", true)) : null);
+  }
+
+  @VisibleForTesting
+  public void start(ScheduledExecutorService executorService) {
+    mExecutorService = executorService;
+  }
+
+  @Override
+  public void beforeShutdown() {
+    if (mExecutorService != null) {
+      flushUpdates();
+    }
+  }
+
+  public void stop() {
+    if (mExecutorService != null) {
+      ThreadUtils.shutdownAndAwaitTermination(mExecutorService, mShutdownTimeout);
+    }
+  }
+
+  /**
+   * Update inode last access time.
+   *
+   * @param context the journal context
+   * @param inode the target inode
+   * @param opTimeMs the last access time to set on the inode
+   */
+  public void updateAccessTime(JournalContext context, Inode inode, long opTimeMs) {
+    if (opTimeMs - inode.getLastAccessTimeMs() > mUpdatePrecision) {
+      try (LockResource lr = mInodeTree.getInodeLockManager().lockUpdate(inode.getId())) {
+        UpdateInodeEntry entry = mInodeTree.updateInodeAccessTime(inode.getId(), opTimeMs);
+        if (mExecutorService != null) {
+          scheduleJournalUpdate(entry);
+        } else {
+          journalAccessTime(context, entry);
+        }
+      }
+    }
+  }
+
+  private void scheduleJournalUpdate(UpdateInodeEntry entry) {
+    mAccessTimeUpdates.add(entry.getId());
+    if (mUpdateScheduled.compareAndSet(false, true)) {
+      mExecutorService.schedule(this::flushScheduledUpdates, mFlushInterval, TimeUnit.MILLISECONDS);
+    }
+  }
+
+  private void journalAccessTime(JournalContext context, UpdateInodeEntry entry) {
+    context.append(Journal.JournalEntry.newBuilder().setUpdateInode(entry)
+        .build());
+  }
+
+  private void flushScheduledUpdates() {
+    mUpdateScheduled.set(false);
+    flushUpdates();
+  }
+
+  private void flushUpdates() {
+    long filesRemoved = 0;
+    try (JournalContext context = mFileSystemMaster.createJournalContext()) {
+      for (Iterator<Long> iterator = mAccessTimeUpdates.iterator();
+           iterator.hasNext();) {
+        long inodeId = iterator.next();
+        iterator.remove();
+        try (LockedInodePath path = mInodeTree.lockFullInodePath(inodeId, LockPattern.READ);
```

Comment:
So, this will re-lock all the inodes? Will this be too expensive? What are the tradeoffs of just journaling without locking?

Reply:
The main risk is that journal entry might be written after an inode is removed, causing the journal replay to fail later on.

Reply:
Is this because we are re-using the `UpdateInode` journal entry? Would it make sense to have a separate `UpdateAccessTime` entry?

Reply:
It looks like we recently deprecated a similar journal entry `UpdateModificationTimeEntry` to unify journaling. I can just ignore missing inode error when applying entries that only update access time.